gemrb/core/Map.cpp

```diff
@@ -2707,243 +2610,177 @@ PathNode* Map::GetLine(const Point &start, const Point &dest, int Speed, int Ori
+									(unsigned) smptChild.x >= Width ||
+									(unsigned) smptChild.y >= Height;
+			bool childBlocked = CheckSearchmapPointFlags(smptChild.x * 16 + 8, smptChild.y * 12 + 6, size,
+												PATH_MAP_PASSABLE, actorsAreBlocking);
+			Actor* childActor = GetActor(NavmapPoint(smptChild.x * 16 + 8, smptChild.y * 12 + 6), GA_NO_DEAD|GA_NO_UNSCHEDULED);
+			bool childIsUnbumpable = childActor && !childActor->ValidTarget(GA_ONLY_BUMPABLE);
+			if (!childOutsideMap && !childBlocked && !isClosed[smptChild.y * Width + smptChild.x] && !childIsUnbumpable) {
+				if (!distFromStart[smptChild.y * Width + smptChild.x] && smptChild != smptSource) {
+					distFromStart[smptChild.y * Width + smptChild.x] = MAX_PATH_COST;
+				}
+				// Theta-star search
+				smptParent = parents[smptCurrent.y * Width + smptCurrent.x];
+				unsigned short adjParentDist = Distance(smptParent, smptChild);
+				unsigned short oldDist = distFromStart[smptChild.y * Width + smptChild.x];
+				unsigned short newDist = distFromStart[smptParent.y * Width + smptParent.x] + adjParentDist;
+				if (newDist < oldDist && newDist < MAX_PATH_COST) {
+					parents[smptChild.y * Width + smptChild.x] = smptParent;
+					distFromStart[smptChild.y * Width + smptChild.x] = newDist;
+					// Calculate heuristic
+					int xDist = smptChild.x - smptDest.x;
+					int yDist = smptChild.y - smptDest.y;
+					// Tie-breaking used to smooth out the path
+					int crossProduct = std::abs(xDist * dyCross - yDist * dxCross);
+					unsigned int heuristic = weight * (Distance(smptChild, smptDest) + (crossProduct >> 9));
+					unsigned int estDist = newDist + heuristic;
+					newNode.point = smptChild;
+					newNode.dist = estDist;
+					open.emplace(newNode);
+				}
+			}
 		}
-
-		Return = new PathNode;
-		Return->Next = StartNode;
-		Return->Next->Parent = Return;
-		StartNode = Return;
-
-		StartNode->x = n.x;
-		StartNode->y = n.y;
-		StartNode->orient = GetOrient( p, n );
-		p = n;
 	}
 
-	return Return;
-}
-
-PathNode* Map::FindPath(const Point &s, const Point &d, unsigned int size, int MinDistance)
-{
-	Point start( s.x/16, s.y/12 );
-	Point goal ( d.x/16, d.y/12 );
-	memset( MapSet, 0, Width * Height * sizeof( unsigned short ) );
-	while (InternalStack.size())
-		InternalStack.pop();
-
-	if (GetBlocked( d.x, d.y, size )) {
-		AdjustPosition( goal );
+	if (foundPath) {
+		return BuildActorPath(smptCurrent, smptDest, parents, backAway);
 	}
-	unsigned int pos = ( goal.x << 16 ) | goal.y;
-	unsigned int pos2 = ( start.x << 16 ) | start.y;
-	InternalStack.push( pos );
-	MapSet[goal.y * Width + goal.x] = 1;
 
-	while (InternalStack.size()) {
-		pos = InternalStack.front();
-		InternalStack.pop();
-		unsigned int x = pos >> 16;
-		unsigned int y = pos & 0xffff;
+	return NULL;
+}
 
-		if (pos == pos2) {
-			//We've found _a_ path
-			//print("GOAL!!!");
-			break;
+PathNode *
+Map::BuildActorPath(SearchmapPoint &smptCurrent, const SearchmapPoint &smptDest, const std::vector<SearchmapPoint> parents,
```

Comment:
Codacy found an issue: [Function parameter 'parents' should be passed by const reference.](https://app.codacy.com/manual/gemrb/gemrb/pullRequest?prid=3427588)